File: Graph/connectingDots/connectingDots.js
(function () {

    let xdir = [1,-1,0,0];
    let ydir = [0,0,1,-1];

    function isSafe(board,rows,cols,i,j,currentChar) {
        if (i>=0 && j>=0 && i<rows && j < cols && board[i][j] === currentChar) return true;
        return false;
    }


    function helper(board, rows,cols, i, j, fromX, fromY, currentChar, visited) {
        if (i > rows || j > cols) return false;

        if (visited[i][j]) return true;

        if (board[i][j] != currentChar) return false;

        visited[i][j] = true;

        for (let k=0; k<4; k++) {
            let nextI = xdir[k];
            let nextJ = ydir[k];
            if(isSafe(board,rows,cols,nextI,nextJ, currentChar) && (nextI != fromX || nextJ != fromY)) {
                if (helper(board,rows,cols,nextI,nextJ,i,j,currentChar,visited)) {
                    return true;
                }
            }
        }

        return false;
    } 


    function hasCycle(board, rows, cols) {
        const visited = [];
        for (let i=0; i<rows; i++) {
            visited[i] = [];
            for (let j=0; j<cols; j++) {
                visited[i][j] = false;
            }
        }

        for (let i=0; i<rows; i++) {
            for (let j=0; j<cols; j++) {
                if (!visited[i][j]) {
                    let currentChar = board[i][j];
                    if (helper(board,rows,cols,i,j,-1,-1,currentChar,visited)) return true;
                }
            }
        }

        return false;
    }



    (function main(){

        let n = 3;
        let m = 4;

        const board = [
            ["A","A","A","A"],
            ["A","B","C","A"],
            ["A","A","A","A"]
        ];


        let ans = hasCycle(board,n,m);
        console.log(ans);

    }());
}());

File: Graph/connectingDots/connectingDots.cpp
#include<iostream>
#include<vector>
using std::cin;
using std::cout;
using std::endl;
using std::vector;
using std::boolalpha;

int xdir[4]{1, -1, 0, 0};
int ydir[4]{0 ,0, 1, -1};

bool isSafe(vector<vector<char> >& board, int rows, int cols, int i, int j, char currentChar) {
    if (i >= 0 && j >= 0 && i < rows && j < cols && board[i][j] == currentChar) return true;
    return false;
}

bool helper(vector<vector<char> >& board, int rows, int cols, int i, int j, int fromX, int fromY, char currentChar, vector<vector<bool> >& visited) {
    if (i > rows || j > cols) return false;
    if (visited[i][j]) return true;
    if (board[i][j] != currentChar) return false;
    visited[i][j] = true;
    for (int k{}; k < 4; k++) {
        int nextI{i + xdir[k]};
        int nextJ{j + ydir[k]};
        if (isSafe(board, rows, cols, nextI, nextJ, currentChar) && (nextI != fromX || nextJ != fromY)) {
            if (helper(board, rows, cols, nextI, nextJ, i, j, currentChar, visited)) return true;
        }
    }
    return false;
}

bool hasCycle(vector<vector<char> >& board, int rows, int cols) {
    vector<vector<bool> > visited(rows, vector<bool>(cols));
    for (int i{}; i < rows; i++) {
        for (int j{}; j < cols; j++) visited[i][j] = false;
    }

    for (int i{}; i < rows; i++) {
        for (int j{}; j < cols; j++) {
            if (!visited[i][j]) {
                char current{board[i][j]};
                if (helper(board, rows, cols, i, j, -1, -1, current, visited)) return true;
            }
        }
    }
    return false;
}

int main() {
    int n, m;
    cin >> n >> m;
    vector<vector<char> > board(n, vector<char>(m));
    for (int i{}; i < n; i++) {
        for (int j{}; j < m; j++) cin >> board[i][j];
    }
    bool ans{hasCycle(board, n, m)};
    cout << boolalpha << ans << endl;
}
